# Forecasting Minimum Daily Temperatures Using Prophet

This project uses Facebook Prophet to forecast daily minimum temperatures and assess the risk of freezing conditions for planting decisions.

We evaluate the next 30 days and use a 95% confidence interval to ensure low-risk recommendations. The analysis converts the output to Fahrenheit and flags days with potential freezing risk to support planting decisions.


In [19]:
import pandas as pd
from prophet import Prophet
import matplotlib.pyplot as plt

# Load sample time-series data
df = pd.read_csv('https://raw.githubusercontent.com/jbrownlee/Datasets/master/daily-min-temperatures.csv')
df.columns = ['ds', 'y']  # Prophet requires these column names

# Fit model
model = Prophet()
model.fit(df)

# Forecast 30 days ahead
future = model.make_future_dataframe(periods=30)
forecast = model.predict(future)

21:30:00 - cmdstanpy - INFO - Chain [1] start processing
21:30:00 - cmdstanpy - INFO - Chain [1] done processing


In [ ]:
## Step 1: Load and Prepare the Data

We’re using a public dataset of daily minimum temperatures in Melbourne, Australia, from 1981–1990. The dataset is renamed to fit Prophet’s expected format (`ds`, `y`).


In [20]:
# Show the structure of the forecast
forecast[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail(15)



ds       yhat  yhat_lower  yhat_upper
3665 1991-01-16  15.721895   12.402685   19.021573
3666 1991-01-17  15.561552   11.764579   18.957908
3667 1991-01-18  15.516950   12.049855   19.012216
3668 1991-01-19  15.458058   11.990418   19.033554
3669 1991-01-20  15.342818   11.797754   18.784317
3670 1991-01-21  15.451486   11.916087   19.101950
3671 1991-01-22  15.523824   12.161705   18.874996
3672 1991-01-23  15.582996   12.019235   18.825175
3673 1991-01-24  15.406452   12.001920   18.762645
3674 1991-01-25  15.354209   11.817531   18.724854
3675 1991-01-26  15.296390   11.967182   18.709161
3676 1991-01-27  15.190848   11.840373   18.549987
3677 1991-01-28  15.317514   11.926225   18.882673
3678 1991-01-29  15.415594   12.293203   19.006773
3679 1991-01-30  15.507480   12.037074   19.020189

In [5]:
model = Prophet(interval_width=0.95)

In [6]:
model.fit(df)
future = model.make_future_dataframe(periods=30)
forecast = model.predict(future)


21:24:16 - cmdstanpy - INFO - Chain [1] start processing
21:24:16 - cmdstanpy - INFO - Chain [1] done processing


In [7]:
forecast[['ds', 'yhat', 'yhat_lower', 'yhat_upper']].tail(10)


ds       yhat  yhat_lower  yhat_upper
3670 1991-01-21  15.451486   10.249149   20.138708
3671 1991-01-22  15.523824   10.195588   21.392787
3672 1991-01-23  15.582996   10.052879   20.808009
3673 1991-01-24  15.406452   10.366000   20.539324
3674 1991-01-25  15.354209   10.015569   20.803046
3675 1991-01-26  15.296390    9.825337   20.527376
3676 1991-01-27  15.190848    9.552032   20.561270
3677 1991-01-28  15.317514   10.112271   20.479363
3678 1991-01-29  15.415594   10.448463   20.815205
3679 1991-01-30  15.507480   10.632307   20.956627

In [23]:
# Add Fahrenheit columns
forecast['yhat_f'] = forecast['yhat'] * 9/5 + 32
forecast['yhat_lower_f'] = forecast['yhat_lower'] * 9/5 + 32
forecast['yhat_upper_f'] = forecast['yhat_upper'] * 9/5 + 32


In [24]:
forecast[['ds', 'yhat_f', 'yhat_lower_f', 'yhat_upper_f']].tail(10)

ds     yhat_f  yhat_lower_f  yhat_upper_f
3670 1991-01-21  59.812676     53.448957     66.383511
3671 1991-01-22  59.942884     53.891068     65.974992
3672 1991-01-23  60.049393     53.634623     65.885314
3673 1991-01-24  59.731614     53.603455     65.772761
3674 1991-01-25  59.637577     53.271556     65.704737
3675 1991-01-26  59.533503     53.540927     65.676490
3676 1991-01-27  59.343526     53.312671     65.389977
3677 1991-01-28  59.571526     53.467205     65.988811
3678 1991-01-29  59.748070     54.127765     66.212191
3679 1991-01-30  59.913465     53.666733     66.236340

In [10]:
# Flag dates where there's a risk of freezing
forecast['freezing_risk'] = forecast['yhat_lower_f'] < 32


In [25]:
forecast['yhat_f'] = forecast['yhat'] * 9/5 + 32
forecast['yhat_lower_f'] = forecast['yhat_lower'] * 9/5 + 32
forecast['yhat_upper_f'] = forecast['yhat_upper'] * 9/5 + 32
forecast['freezing_risk'] = forecast['yhat_lower_f'] < 32


In [26]:
forecast[['ds', 'yhat_f', 'yhat_lower_f', 'yhat_upper_f', 'freezing_risk']].tail(15)


ds     yhat_f  yhat_lower_f  yhat_upper_f  freezing_risk
3665 1991-01-16  60.299411     54.324833     66.238832          False
3666 1991-01-17  60.010794     53.176242     66.124234          False
3667 1991-01-18  59.930509     53.689739     66.221989          False
3668 1991-01-19  59.824505     53.582752     66.260396          False
3669 1991-01-20  59.617072     53.235958     65.811770          False
3670 1991-01-21  59.812676     53.448957     66.383511          False
3671 1991-01-22  59.942884     53.891068     65.974992          False
3672 1991-01-23  60.049393     53.634623     65.885314          False
3673 1991-01-24  59.731614     53.603455     65.772761          False
3674 1991-01-25  59.637577     53.271556     65.704737          False
3675 1991-01-26  59.533503     53.540927     65.676490          False
3676 1991-01-27  59.343526     53.312671     65.389977          False
3677 1991-01-28  59.571526     53.467205     65.988811          False
3678 1991-01-29  59.748070     54.127765     66.212191          False
3679 1991-01-30  59.913465     53.666733     66.236340          False

## Conclusion

Based on the 30-day forecast generated by the Prophet model (using a 95% confidence interval), **no dates were flagged as having a freezing risk**. All forecasted lower-bound temperatures in Fahrenheit remained above 32°F.

This suggests that **minimum temperature is not a limiting factor** for planting during this time period. Most common plants that are sensitive to frost or freezing conditions can likely be planted without risk, assuming no unexpected cold fronts occur outside the model's forecast.

While other environmental factors (e.g., soil moisture, sunlight, wind) may still influence planting success, the temperature data supports a decision to proceed with planting during this window.


 ## Understanding the Forecast Output

The Prophet model outputs several key columns for each forecasted date:

- **`yhat`**: The predicted minimum temperature for that date.
- **`yhat_lower`**: The lower bound of the confidence interval.
- **`yhat_upper`**: The upper bound of the confidence interval.

Since we set the model's `interval_width` to `0.95`, this means:

> We are **95% confident** that the actual minimum temperature will fall between `yhat_lower` and `yhat_upper`.

This confidence interval reflects the uncertainty in the forecast. A **wider interval** indicates more uncertainty, while a **narrower interval** reflects higher certainty (but more risk of being wrong).

### Example:

If the forecast for January 30, 1991 shows:

- `yhat = 15.5°C`
- `yhat_lower = 12.0°C`
- `yhat_upper = 18.9°C`

We interpret this as:

> “The model forecasts the minimum temperature to be around **15.5°C**, with a **95% chance** it falls between **12.0°C and 18.9°C**.”

This logic also applies to the converted Fahrenheit values (`yhat_f`, `yhat_lower_f`, `yhat_upper_f`), which support real-world decisions (e.g., whether to plant based on frost risk).
